File: src/frontend/src/types/planner.ts
/**
 * The planner's wire contract (SSE v2, TRA-142) as the frontend consumes it.
 *
 * `ai_api` defines these as Pydantic models and exports them through
 * `just contracts` into `types/generated/ai-api.ts` (never edit that file by
 * hand; re-run `just contracts` after a backend schema change). This module
 * re-exports the generated schemas under the names the planner UI (TRA-144)
 * already uses, and derives the small view-model helpers from them.
 */

import type { components } from "@/types/generated/ai-api";

/** A chat turn as `ai_api` replays it; the same shape `/ai/chat` uses. */
export type ChatMessage = components["schemas"]["ChatMessage"];

/** A city the planner covers (`GET /ai/planner/cities`): a destination to offer. */
export type PlannerCity = components["schemas"]["PlannerCity"];

/** A city's own description in one language, with the page it comes from (TRA-182). */
export type CityIntro = components["schemas"]["CityIntro"];

// ─── The brief ──────────────────────────────────────────────────────────────

export type TripBrief = components["schemas"]["TripBrief"];

/** The five checklist fields the brief must have before a trip is generated. */
export type BriefField = components["schemas"]["BriefEvent"]["missing"][number];

export const BRIEF_FIELDS = [
  "destination",
  "origin",
  "dates",
  "travellers",
  "interests",
] as const satisfies readonly BriefField[];

/** `€`, `€€`, `€€€`: the only way a price is ever shown (never a number). */
export type PriceTier = NonNullable<components["schemas"]["OptionCard"]["price_tier"]>;

export type Pace = NonNullable<TripBrief["pace"]>;

export const EMPTY_BRIEF: TripBrief = {
  destination: null,
  origin: null,
  start_date: null,
  end_date: null,
  nights: null,
  adults: null,
  children: null,
  budget_tier: null,
  interests: [],
  pace: null,
};

// ─── Slots and options ────────────────────────────────────────────────────────

export type Slot = components["schemas"]["Slot"];

export type DayPart = NonNullable<Slot["part"]>;

export const DAY_PARTS = ["morning", "afternoon", "evening", "night"] as const satisfies readonly DayPart[];

/**
 * One selectable card, hydrated by the backend from the corpus document
 * `id` refers to. The model only picks ids and writes `why`.
 */
export type OptionCard = components["schemas"]["OptionCard"];

/**
 * The same card in full (`GET /ai/planner/card?id=`), as the detail panel
 * shows it: the corpus article behind the card, its address, phone and site.
 *
 * Additive over {@link OptionCard} in shape, but not a replacement for one:
 * `why` is what the model wrote for a turn, so the endpoint returns `""`, and
 * `image_url`/`image_credit` hold the corpus's own photo or one found on
 * Wikimedia Commons — `null` when neither has one, where the streamed card
 * carries a fallback picture. So merge the detail onto the card already held
 * instead of replacing it — keep that card's `why`, and its photo and credit
 * together when the detail brings none:
 * `{ ...card, ...detail, why: card.why, ...(detail.image_url ? {} : { image_url: card.image_url, image_credit: card.image_credit }) }`
 */
export type CardDetail = components["schemas"]["CardDetail"];

export type OptionKind = components["schemas"]["OptionsEvent"]["kind"];

export const OPTION_KINDS = [
  "neighbourhood",
  "hotel",
  "experience",
  "restaurant",
  "flight",
  "day_template",
] as const satisfies readonly OptionKind[];

export type SelectionMode = components["schemas"]["OptionsEvent"]["selection"];

/** A carousel of options the user can pick from; `group_id` names it in the next turn. */
export type OptionsGroup = Omit<components["schemas"]["OptionsEvent"], "type">;

// ─── Itinerary ops ────────────────────────────────────────────────────────────

export type WarnCode = components["schemas"]["WarnOp"]["code"];

export const WARN_CODES = [
  "too_far",
  "closed",
  "overloaded_day",
  "unverified_price",
] as const satisfies readonly WarnCode[];

export type RouteInfo = Omit<components["schemas"]["SetRouteOp"], "op">;

export type DayWeather = Omit<components["schemas"]["SetWeatherOp"], "op" | "day">;

/** Ops in an `itinerary_patch`, discriminated on `op`. */
export type ItineraryOp = components["schemas"]["ItineraryOp"];

// ─── Events ─────────────────────────────────────────────────────────────────

export type PlannerEvent = components["schemas"]["PlannerEvent"];

export type PlannerEventType = PlannerEvent["type"];

// ─── The request ──────────────────────────────────────────────────────────────

export type SelectAction = components["schemas"]["SelectAction"];

export type RemoveAction = components["schemas"]["RemoveAction"];

/** Days → parts → card ids, plus the stay: what the client has, so `ai_api` stays stateless. */
export type ItinerarySnapshot = components["schemas"]["ItinerarySnapshot"];

/** One turn as the app sends it: every key present, `null` when unknown. */
export type PlannerTurn = components["schemas"]["PlannerTurn"];

File: src/backend/services/ai_api/ai_api/schemas/planner.py
"""The planner's request: one turn of the conversation (TRA-142).

`ai_api` keeps no state between requests, so the client sends what the
planner has to know each time: the transcript, the brief and the itinerary it
holds (card ids only; the cards themselves live in the corpus). Limits follow
`schemas/chat.py`. Like the events, nothing here has a default: the page sends
every field (`null` when unknown), and the generated type says so.
"""

from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, Field

from ai_api.domain.models import City
from ai_api.schemas.chat import MAX_HISTORY_TURNS, MAX_MESSAGE_CHARS, ChatMessage
from ai_api.schemas.planner_events import OptionCard, Slot, TripBrief

MAX_CARD_IDS = 20
"""Most ids a selection or a slot may carry."""

MAX_CARD_ID_CHARS = 200
"""A corpus id is a short key (`osm:relation/13067`), never a payload."""

MAX_EXCLUDED_CARD_IDS = 60
"""Most ids one ask may rule out: `OPTIONS_COUNT` a page, a few pages deep."""

MAX_DAYS = 14
"""Longest itinerary a snapshot may describe."""


CardId = Annotated[str, Field(min_length=1, max_length=MAX_CARD_ID_CHARS)]
"""One corpus document id, as a request carries it."""


class SelectAction(BaseModel):
    """Cards chosen in a carousel the server sent earlier (`group_id`).

    A group whose id names a slot (`slot:<day>:<part>`) places the pick on
    its own. A `found:` group does not: its cards arrived unplaced, so the
    page names the day and the part the traveller chose in `slot` (TRA-185,
    ADR 0018). `null` for every placed group.
    """

    type: Literal["select"]
    group_id: str = Field(min_length=1, max_length=120)
    card_ids: list[str] = Field(min_length=1, max_length=MAX_CARD_IDS)
    slot: Slot | None


class RemoveAction(BaseModel):
    """An activity taken out of a slot."""

    type: Literal["remove"]
    slot: Slot
    card_id: str = Field(min_length=1)


PlannerAction = Annotated[SelectAction | RemoveAction, Field(discriminator="type")]


class DaySlots(BaseModel):
    """Card ids per part of the day. Every part is sent, empty or not, so the
    generated type matches the page's `Record<DayPart, string[]>`."""

    morning: list[str] = Field(max_length=MAX_CARD_IDS)
    afternoon: list[str] = Field(max_length=MAX_CARD_IDS)
    evening: list[str] = Field(max_length=MAX_CARD_IDS)
    night: list[str] = Field(max_length=MAX_CARD_IDS)


class DaySnapshot(BaseModel):
    day: int = Field(ge=1)
    slots: DaySlots


class ItinerarySnapshot(BaseModel):
    """What the client holds: the stay and the days, as card ids."""

    stay_card_id: str | None
    days: list[DaySnapshot] = Field(max_length=MAX_DAYS)


class PlannerTurn(BaseModel):
    """One turn: a message, a structured action, or both."""

    message: str | None = Field(max_length=MAX_MESSAGE_CHARS)
    action: PlannerAction | None
    history: list[ChatMessage] = Field(
        max_length=MAX_HISTORY_TURNS,
        description="Text turns before this one, oldest first",
    )
    brief: TripBrief | None
    itinerary: ItinerarySnapshot | None
    exclude_card_ids: list[CardId] = Field(
        max_length=MAX_EXCLUDED_CARD_IDS,
        description="ids the traveller already saw for this ask; never offered again",
    )
    trip_id: UUID | None = Field(
        description="Reserved: the Trip this draft will be saved to"
    )


class CardDetail(OptionCard):
    """One card in full (`GET /planner/card?id=`): what the panel shows when a
    traveller opens an activity, on top of everything the carousel card has.

    Additive over `OptionCard` and read from the same place — the corpus
    document the id names — so the client never sends content, only an id.
    Like every planner model, no field has a default: an unknown one travels
    as `null` and the generated TypeScript keeps it required.
    """

    description: str = Field(
        description=(
            "The corpus document's own text, trimmed at a sentence boundary; "
            "empty when the document carries none"
        )
    )
    address: str | None
    phone: str | None
    website: str | None = Field(
        description="The venue's own site, when it is not the source page"
    )
    heading_path: str | None = Field(
        description="Where the document sits in the city's corpus"
    )


class CityIntro(BaseModel):
    """A city's own description in one language: the lead of its Wikivoyage
    article (CC BY-SA 4.0), trimmed at a sentence boundary, and the page it
    comes from — shown with the text, as the licence asks."""

    text: str
    source_url: str


class PlannerCity(BaseModel):
    """A city the planner covers (`GET /planner/cities`): what the page needs
    to offer it as a destination and to introduce it on a trip overview. The
    spellings it answers to stay server-side."""

    slug: str
    name: str
    centre: tuple[float, float] = Field(description="[latitude, longitude]")
    timezone: str
    intro: dict[str, CityIntro] = Field(
        description=(
            "The city's description keyed by language code (`en`, `es`); a "
            "language the corpus has no lead for is absent"
        )
    )
    image_url: str | None = Field(
        description="The city's photo on Wikimedia Commons, 1200 px wide"
    )
    image_credit: str | None = Field(
        description="The line to print beside the photo (author and licence)"
    )


def planner_city(city: City) -> PlannerCity:
    return PlannerCity(
        slug=city.slug,
        name=city.name,
        centre=city.centre,
        timezone=city.timezone,
        intro={
            lang: CityIntro(text=intro.text, source_url=intro.source_url)
            for lang, intro in city.intro.items()
        },
        image_url=city.image_url,
        image_credit=city.image_credit,
    )
